ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,173 @@
+        // MerkeTree.verify(allParachainHeadsMerkleTreeRoot, ourParachainMerkleProof)
+        // returns parachainBlockHeader
+
+        // Prove that the commitment is in fact in the parachain block header
+        // require(parachainBlockHeader.commitment == commitment)
+
+        // Validate that the commitment matches the commitment contents
+        require(
+            validateCommitment(commitment, commitmentContents),
+            "invalid commitment"
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >=
+                commitmentContents.messages.length * MAX_GAS_PER_MESSAGE,
+            "insufficient gas for delivery of all messages"
+        );
+
+        // Require all payloads are smaller than max_payload_size
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            require(
+                commitmentContents.messages[i].payload.length <=
+                    MAX_PAYLOAD_BYTE_SIZE,
+                "message payload bytesize exceeds maximum payload size"
+            );
+        }
+        return true;
+    }
+
+    function processCommitmentContents(
+        CommitmentContents memory commitmentContents
+    ) internal {
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            Message memory message = commitmentContents.messages[i];
+            require(message.nonce == lastProcessedNonce + 1, "invalid nonce");
+
+            lastProcessedNonce = lastProcessedNonce + 1;
+
+            // Deliver the message to the destination
+            // Delivery will have fixed maximum gas allowed for the destination app.
+            address targetApplicationAddress = message.targetApplicationAddress;
+            uint256 allowedGas = MAX_GAS_PER_MESSAGE;
+            bytes memory callInput = message.payload;
+
+            bool result;
+            uint256 callInputLength = callInput.length;
+            assembly {
+                let callInputDataPointer := add(callInput, 32) // Move 32 ahead of callInput length slot to get to actual data slot (TODO: verify this works in all cases)
+
+                // Dispatch call to the receiver - it is expected to be fire and forget. If the call reverts, runs out of gas, error,
+                // etc, its the fault of the sender
+                result := call(
+                    allowedGas, // Allowed gas
+                    targetApplicationAddress, // To addr
+                    0, // No ether value being sent
+                    callInputDataPointer, // Inputs are stored at callInputDataPointer
+                    callInputLength, // Input length
+                    callInputDataPointer, // Store output over input (saves space)
+                    callInputLength // Outputs can be as long as input length
+                )
```

Comment:
Can we add a task to think about security vulnerabilities of this mechanism? Re-entrancy into the receive channel does not seem to be possible (a message would need to know the commitment hash for re-entrancy), but we should look into other potential attacks.

Reply:
sounds good - will be simpler too if we replace this with a solidity call rather than a assembly call

Reply:
Addressed in https://github.com/Snowfork/polkadot-ethereum/pull/219/commits/c1c84636367643bd86fe7319869912a7dc2cdb2d